[BOLT] Simplify RAState helpers (NFCI)

bgergely0 opened the pull request:
- unify isRAStateSigned and isRAStateUnsigned to a common getRAState,
- unify setRASigned and setRAUnsigned into setRAState(MCInst, bool),
- update users of these to match the new implementations.

bgergely0 committed:
[BOLT] Change RAState helpers (NFCI)

- unify isRAStateSigned and isRAStateUnsigned to a common getRAState,
- unify setRASigned and setRAUnsigned into setRAState(MCInst, bool),
- update users of these to match the new implementations.

peterwaller-arm commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -40,25 +40,30 @@ void InsertNegateRAState::runOnFunction(BinaryFunction &BF) {
     coverFunctionFragmentStart(BF, FF);
     bool FirstIter = true;
     MCInst PrevInst;
+    bool PrevRAState = false;
```
PrevInst now looks unused.

peterwaller-arm commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -81,10 +86,15 @@ void InsertNegateRAState::coverFunctionFragmentStart(BinaryFunction &BF,
+  if (!RAState) {
+    BC.errs() << "BOLT-ERROR: unknown RAState after inferUnknownStates "
+              << " in function " << BF.getPrintName() << "\n";
+    return;
```
From https://llvm.org/docs/CodingStandards.html:

> The ``llvm_unreachable`` function can be used to document areas of control flow
> that should never be entered if the program invariants hold:
> 
> .. code-block:: c++
> 
>   enum { Foo, Bar, Baz } X = foo();
> 
>   switch (X) {
>     case Foo: /* Handle Foo */; break;
>     case Bar: /* Handle Bar */; break;
>     default:
>       llvm_unreachable("X should be Foo or Bar here");
>   }
> 
> Additionally, ``reportFatalInternalError`` can be used to report invariant
> violations even in builds that do not enable assertions:
> 
> .. code-block:: c++
> 
>   if (VerifyFooAnalysis && !Foo.verify()) {
>     reportFatalInternalError("Analysis 'foo' not preserved");
>   }

The coding standards suggest llvm_unreachable may only be visible in assertions builds. If there is a risk this could happen in the real world, a fatal internal error would be preferable. Bolt doesn't use this function, the standard appears to be to simply exit if there is nothing more to be done.

bgergely0 committed:
[BOLT] Review changes

- remove unused PrevInst
- add a static PassFailed, and createFatalBOLTError() if the pass failed
  on any of the threads executed in parallel. This is the same way
  ADRRelaxationPass handles errors in threads.

bgergely0 commented:
Note: the way the errors are reported here does not have to be perfect, as the next PR (#163381) removes these and adds new errors.

paschalis-mpeis reviewed (approved):
Looks good, thanks for the simplification Gergely.

> Note: the way the errors are reported here does not have to be perfect, as the next PR (https://github.com/llvm/llvm-project/pull/163381) removes these and adds new errors.

👍, I may have some follow-up comments there too, on the users (ie insert-negate pass)

bgergely0 committed:
[BOLT] Bugfix: use getFirstNonPseudo() instead of begin()

RA states are only assigned to non-pseudo instructions. Because of this,
CFI instructions are skipped.
Here, I incorrectly used begin(), not skipping CFIs.

paschalis-mpeis reviewed (approved):
Thanks for commit d1f66f71d1ee2de7098faf96c3c4eb63953237c5 that fixes the reported issue. Accepting this NFCI simplification as a means of unblocking the stack.

No strong opinion around naming; one suggestion could be `setRASigned(bool)`, to set signed or unsigned (and another clear function in case that's needed).
